{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "15a758ed-5ce4-4617-9683-b96d64cd4d3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import panhumanpy as ph\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "import numpy as np\n",
    "import pooch\n",
    "from scipy.sparse import csr_matrix\n",
    "from scipy.io import mmwrite\n",
    "import re\n",
    "import anndata\n",
    "import scanpy.external as sce\n",
    "import scvi\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import gseapy as gp\n",
    "from gseapy.plot import dotplot\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bf194d31-4ab6-4f50-9690-a98e43a07f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "KPMP_SN_raw=sc.read('/scratch/yongqunh_root/yongqunh0/wruopeng/HuBMAP_KPMP/kpmp-sn-raw-rnaseq.h5ad')\n",
    "HuBMAP_LK_raw=sc.read('/scratch/yongqunh_root/yongqunh0/wruopeng/HuBMAP_KPMP/LK_raw_updated_2025_june.h5ad')\n",
    "HuBMAP_RK_raw=sc.read('/scratch/yongqunh_root/yongqunh0/wruopeng/HuBMAP_KPMP/RK_raw_updated_2025_june.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a85c0de5-30c0-40b4-90ce-fe8743519d4a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['library_id', 'nCount_RNA', 'nFeature_RNA', 'percent.er', 'percent.mt', 'subclass.l2', 'subclass.l1', 'class', 'experiment_id', 'suspension_type', 'assay_ontology_term_id', 'organism_ontology_term_id', 'donor_id', 'specimen', 'disease_category', 'disease_ontology_term_id', 'eGFR', 'diabetes_history', 'hypertension', 'sex_ontology_term_id', 'development_stage_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'is_primary_data', 'region', 'percent.cortex', 'percent.medulla', 'tissue_type', 'tissue_ontology_term_id', 'cell_type_ontology_term_id', 'Age_binned', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid']\n",
      "['kidney interstitial alternatively activated macrophage', 'kidney distal convoluted tubule epithelial cell', 'epithelial cell of proximal tubule', 'T cell', 'kidney loop of Henle thick ascending limb epithelial cell', 'kidney collecting duct intercalated cell', 'kidney interstitial fibroblast', 'kidney connecting tubule epithelial cell', 'kidney collecting duct principal cell', 'endothelial cell', 'B cell', 'mature NK T cell', 'mononuclear phagocyte', 'podocyte', 'parietal epithelial cell', 'mast cell', 'plasmacytoid dendritic cell, human', 'plasma cell', 'non-classical monocyte', 'conventional dendritic cell', 'kidney loop of Henle thin descending limb epithelial cell', 'renal interstitial pericyte', 'kidney loop of Henle thin ascending limb epithelial cell', 'neutrophil', 'papillary tips cell', 'neural cell']\n",
      "['cortex of kidney', 'renal papilla', 'renal medulla', 'kidney']\n",
      "['UBERON:0001225', 'UBERON:0001228', 'UBERON:0000362', 'UBERON:0002113']\n"
     ]
    }
   ],
   "source": [
    "print(list(KPMP_SN_raw.obs.columns))\n",
    "print(list(KPMP_SN_raw.obs['cell_type'].unique()))\n",
    "print(list(KPMP_SN_raw.obs['tissue'].unique()))\n",
    "print(list(KPMP_SN_raw.obs['tissue_ontology_term_id'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5faa2744-9ead-44c2-9e7f-b6851c203d86",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['uuid', 'hubmap_id', 'age', 'sex', 'height', 'weight', 'bmi', 'cause_of_death', 'race', 'barcode', 'dataset', 'azimuth_label', 'azimuth_id', 'predicted_CLID', 'predicted_label', 'cl_match_type', 'prediction_score']\n",
      "['kidney cortex collecting duct principal cell', 'kidney connecting tubule epithelial cell', 'vasa recta descending limb cell', 'kidney inner medulla collecting duct principal cell', 'papillary tips cell', 'renal medullary fibroblast', 'parietal epithelial cell', 'vasa recta ascending limb cell', 'kidney loop of Henle short descending thin limb epithelial cell', 'kidney loop of Henle thin ascending limb epithelial cell', 'kidney interstitial fibroblast', 'kidney resident macrophage', 'kidney loop of Henle long descending thin limb inner medulla epithelial cell', 'kidney loop of Henle medullary thick ascending limb epithelial cell', 'renal interstitial pericyte', 'epithelial cell of proximal tubule segment 3', 'mesangial cell', 'glomerular capillary endothelial cell', 'epithelial cell of early distal convoluted tubule', 'endothelial cell of arteriole', 'macula densa epithelial cell', 'kidney loop of Henle long descending thin limb outer medulla epithelial cell', 'kidney loop of Henle cortical thick ascending limb epithelial cell', 'peritubular capillary endothelial cell', 'kidney outer medulla collecting duct intercalated cell', 'epithelial cell of proximal tubule segment 1', 'non-classical monocyte', 'kidney cortex collecting duct intercalated cell', 'monocyte-derived dendritic cell', 'kidney connecting tubule principal cell', 'mature NK T cell', 'kidney outer medulla collecting duct principal cell', 'T cell', 'neutrophil', 'mononuclear phagocyte', 'renal beta-intercalated cell', 'endothelial cell of lymphatic vessel', nan, 'podocyte', 'B cell', 'Schwann cell', 'plasma cell', 'plasmacytoid dendritic cell, human', 'mast cell', 'conventional dendritic cell', 'epithelial cell of late distal convoluted tubule', 'epithelial cell of proximal tubule segment 2', 'kidney granular cell', 'kidney connecting tubule alpha-intercalated cell']\n",
      "['Cortical Collecting Duct Principal', 'Connecting Tubule', 'Descending Vasa Recta Endothelial ', 'Inner Medullary Collecting Duct', 'Papillary Tip Epithelial', 'Medullary Fibroblast', 'Parietal Epithelial', 'Ascending Vasa Recta Endothelial', 'Descending Thin Limb Type 1', 'Ascending Thin Limb', 'Fibroblast', 'M2 Macrophage', 'Descending Thin Limb Type 3', 'Medullary Thick Ascending Limb', 'Vascular Smooth Muscle / Pericyte', 'Proximal Tubule Epithelial Segment 3', 'Mesangial', 'Glomerular Capillary Endothelial', 'Distal Convoluted Tubule Type 1', 'Afferent / Efferent Arteriole Endothelial', 'Macula Densa', 'Descending Thin Limb Type 2', 'Cortical Thick Ascending Limb', 'Peritubular Capilary Endothelial ', 'Outer Medullary Collecting Duct Intercalated Type A', 'Proximal Tubule Epithelial Segment 1', 'Non-classical monocyte', 'Cortical Vascular Smooth Muscle / Pericyte', 'Cortical Collecting Duct Intercalated Type A', 'Monocyte-derived', 'Connecting Tubule Principal', 'Natural Killer T', 'Outer Medullary Collecting Duct Principal', 'T', 'Neutrophil', 'Cycling Mononuclear Phagocyte', 'Intercalated Type B', 'Lymphatic Endothelial', nan, 'Podocyte', 'B', 'Schwann / Neural', 'Plasma', 'Plasmacytoid Dendritic', 'Mast', 'Classical Dendritic', 'Distal Convoluted Tubule Type 2', 'Proximal Tubule Epithelial Segment 2', 'Renin-positive Juxtaglomerular Granular', 'Connecting Tubule Intercalated Type A']\n"
     ]
    }
   ],
   "source": [
    "print(list(HuBMAP_RK_raw.obs.columns))\n",
    "print(list(HuBMAP_RK_raw.obs['predicted_label'].unique()))\n",
    "print(list(HuBMAP_RK_raw.obs['azimuth_label'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f66b27bc-b8d6-48ae-b298-868cb3e3b200",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "67f1dfc3-e3e7-488b-acbb-0258af8fc79f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['feature_name', 'feature_reference', 'feature_biotype',\n",
       "       'feature_length'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KPMP_SN_raw.var.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ae884cea-c505-4612-868a-77d9c01b542f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"ENSG00000272954.1\" in KPMP_SN_raw.var[\"feature_name\"].astype(str).values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "05553583-4d86-40ae-980b-1303ca23421a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered to dataset cell types (predicted_label): kept 1371041/1522897 cells (151856 removed).\n",
      "Filtered to dataset cell types (predicted_label): kept 1673578/1803736 cells (130158 removed).\n",
      "Removing slide_seq cells only:\n",
      "  Cells before: 1371041\n",
      "  Cells after:  960932\n",
      "  Cells removed: 410109\n",
      "Removing slide_seq cells only:\n",
      "  Cells before: 1673578\n",
      "  Cells after:  1283320\n",
      "  Cells removed: 390258\n",
      "Identified and updated 36297 common gene IDs.\n",
      "Identified and updated 36297 common gene IDs.\n",
      "Found 36297 genes shared across all datasets.\n",
      "Obs columns before cocnat: ['consortium', 'collection', 'dataset_id', 'as_id', 'cl_id', 'cl_label', 'age', 'sex', 'race', 'disease']\n",
      "Obs columns before cocnat: ['consortium', 'collection', 'dataset_id', 'as_id', 'cl_id', 'cl_label', 'age', 'sex', 'race', 'disease']\n",
      "Obs columns before cocnat: ['consortium', 'collection', 'dataset_id', 'as_id', 'cl_id', 'cl_label', 'age', 'sex', 'race', 'disease']\n"
     ]
    }
   ],
   "source": [
    "KPMP_shared = {\n",
    "    \"library_id\": \"dataset_id\",\n",
    "    \"tissue_ontology_term_id\": \"as_id\",\n",
    "    \"cell_type_ontology_term_id\": \"cl_id\",\n",
    "    \"cell_type\": \"cl_label\",\n",
    "    \"nCount_RNA\": \"gene_count\",\n",
    "    \"Age_binned\": \"age\",\n",
    "    \"sex\": \"sex\",\n",
    "    \"self_reported_ethnicity\": \"race\",\n",
    "    \"disease\": \"disease\",\n",
    "}\n",
    "KPMP_SN_raw.obs['consortium'] = 'KPMP'\n",
    "KPMP_SN_raw.obs['collection'] = 'KPMP SC RNAseq'\n",
    "\n",
    "def filter_to_dataset_cell_types(\n",
    "    adata,\n",
    "    dataset_names,\n",
    "    source_col='cell_type'\n",
    "):\n",
    "    # count before\n",
    "    before = adata.n_obs\n",
    "\n",
    "    # filter mask\n",
    "    mask = adata.obs[source_col].isin(dataset_names)\n",
    "    adata_filtered = adata[mask, :].copy()\n",
    "\n",
    "    # report\n",
    "    after = adata_filtered.n_obs\n",
    "    print(\n",
    "        f\"Filtered to dataset cell types ({source_col}): \"\n",
    "        f\"kept {after}/{before} cells ({before - after} removed).\"\n",
    "    )\n",
    "\n",
    "    return adata_filtered\n",
    "\n",
    "mapping = {\n",
    "    # exact matches\n",
    "    'B cell': 'B cell',\n",
    "    'T cell': 'T cell',\n",
    "    'conventional dendritic cell': 'conventional dendritic cell',\n",
    "    'mast cell': 'mast cell',\n",
    "    'mature NK T cell': 'mature NK T cell',\n",
    "    'mononuclear phagocyte': 'mononuclear phagocyte',\n",
    "    'neutrophil': 'neutrophil',\n",
    "    'non-classical monocyte': 'non-classical monocyte',\n",
    "    'papillary tips cell': 'papillary tips cell',\n",
    "    'parietal epithelial cell': 'parietal epithelial cell',\n",
    "    'plasma cell': 'plasma cell',\n",
    "    'plasmacytoid dendritic cell, human': 'plasmacytoid dendritic cell, human',\n",
    "    'podocyte': 'podocyte',\n",
    "    'kidney interstitial fibroblast': 'kidney interstitial fibroblast',\n",
    "    'renal interstitial pericyte': 'renal interstitial pericyte',\n",
    "    # proximal tubule\n",
    "    'epithelial cell of proximal tubule segment 1': 'epithelial cell of proximal tubule',\n",
    "    'epithelial cell of proximal tubule segment 2': 'epithelial cell of proximal tubule',\n",
    "    'epithelial cell of proximal tubule segment 3': 'epithelial cell of proximal tubule',\n",
    "    # distal convoluted tubule\n",
    "    'epithelial cell of early distal convoluted tubule': 'kidney distal convoluted tubule epithelial cell',\n",
    "    'epithelial cell of late distal convoluted tubule': 'kidney distal convoluted tubule epithelial cell',\n",
    "    # connecting tubule & collecting duct\n",
    "    'kidney connecting tubule alpha-intercalated cell': 'kidney collecting duct intercalated cell',\n",
    "    'kidney connecting tubule principal cell': 'kidney collecting duct principal cell',\n",
    "    'kidney cortex collecting duct intercalated cell': 'kidney collecting duct intercalated cell',\n",
    "    'kidney outer medulla collecting duct intercalated cell': 'kidney collecting duct intercalated cell',\n",
    "    'kidney cortex collecting duct principal cell': 'kidney collecting duct principal cell',\n",
    "    'kidney inner medulla collecting duct principal cell': 'kidney collecting duct principal cell',\n",
    "    'kidney outer medulla collecting duct principal cell': 'kidney collecting duct principal cell',\n",
    "    'kidney collecting duct intercalated cell': 'kidney collecting duct intercalated cell',\n",
    "    'kidney collecting duct principal cell': 'kidney collecting duct principal cell',\n",
    "    'kidney connecting tubule epithelial cell': 'kidney connecting tubule epithelial cell',\n",
    "    # loop of Henle\n",
    "    'kidney loop of Henle cortical thick ascending limb epithelial cell': \n",
    "        'kidney loop of Henle thick ascending limb epithelial cell',\n",
    "    'kidney loop of Henle medullary thick ascending limb epithelial cell': \n",
    "        'kidney loop of Henle thick ascending limb epithelial cell',\n",
    "    'kidney loop of Henle thick ascending limb epithelial cell': \n",
    "        'kidney loop of Henle thick ascending limb epithelial cell',\n",
    "    'kidney loop of Henle thin ascending limb epithelial cell': \n",
    "        'kidney loop of Henle thin ascending limb epithelial cell',\n",
    "    'kidney loop of Henle short descending thin limb epithelial cell': \n",
    "        'kidney loop of Henle thin descending limb epithelial cell',\n",
    "    'kidney loop of Henle long descending thin limb inner medulla epithelial cell': \n",
    "        'kidney loop of Henle thin descending limb epithelial cell',\n",
    "    'kidney loop of Henle long descending thin limb outer medulla epithelial cell': \n",
    "        'kidney loop of Henle thin descending limb epithelial cell',\n",
    "    # endothelial\n",
    "    'endothelial cell': 'endothelial cell',\n",
    "}\n",
    "HuBMAP_LK_raw = filter_to_dataset_cell_types(HuBMAP_LK_raw, mapping, source_col='predicted_label')\n",
    "\n",
    "\n",
    "HuBMAP_RK_raw = filter_to_dataset_cell_types(HuBMAP_RK_raw, mapping, source_col='predicted_label')\n",
    "\n",
    "\n",
    "\n",
    "def remove_slide_seq_cells(adata, slide_seq_ids):\n",
    "\n",
    "    # Count before filtering\n",
    "    total_before = adata.n_obs\n",
    "    \n",
    "    # Mask of cells to remove\n",
    "    remove_mask = adata.obs['hubmap_id'].isin(slide_seq_ids)\n",
    "    \n",
    "    # Subset to keep only non-slide_seq cells\n",
    "    adata_filtered = adata[~remove_mask, :].copy()\n",
    "    total_after = adata_filtered.n_obs\n",
    "    \n",
    "    # Report\n",
    "    print(\"Removing slide_seq cells only:\")\n",
    "    print(f\"  Cells before: {total_before}\")\n",
    "    print(f\"  Cells after:  {total_after}\")\n",
    "    print(f\"  Cells removed: {total_before - total_after}\")\n",
    "    return adata_filtered\n",
    "need_to_removed_slide_seq_HuBMAP=['HBM222.VQSW.335',\n",
    " 'HBM248.HPXX.584',\n",
    " 'HBM269.GDLH.894',\n",
    " 'HBM294.XZLM.256',\n",
    " 'HBM356.MDPN.792',\n",
    " 'HBM363.FVKP.935',\n",
    " 'HBM398.BLRW.228',\n",
    " 'HBM456.CGDP.395',\n",
    " 'HBM462.XQCR.933',\n",
    " 'HBM522.QXVG.468',\n",
    " 'HBM528.KNCB.488',\n",
    " 'HBM547.SJSK.268',\n",
    " 'HBM547.TFRR.794',\n",
    " 'HBM595.LBXP.486',\n",
    " 'HBM679.RLJH.432',\n",
    " 'HBM735.FSBZ.626',\n",
    " 'HBM757.KLKW.524',\n",
    " 'HBM775.CMGG.464',\n",
    " 'HBM785.XFTT.663',\n",
    " 'HBM834.SLQN.292',\n",
    " 'HBM883.PHQS.523',\n",
    " 'HBM976.*','HBM232.MBNR.586', 'HBM266.FTJN.632', 'HBM297.FDTX.382', 'HBM445.HBRO.488',\n",
    " 'HBM459.KCST.593', 'HBM532.KKRC.477', 'HBM634.JHVB.286', 'HBM634.ZSHF.736',\n",
    " 'HBM647.QDBG.936', 'HBM736.MNMD.453', 'HBM827.MJMM.447', 'HBM846.KVCF.674',\n",
    " 'HBM892.CCDZ.345', 'HBM965.PSNC.855', 'HBM986.KFWG.239', 'HBM232.MBNR.586']\n",
    "HuBMAP_LK_raw=remove_slide_seq_cells(HuBMAP_LK_raw,need_to_removed_slide_seq_HuBMAP)\n",
    "HuBMAP_RK_raw=remove_slide_seq_cells(HuBMAP_RK_raw,need_to_removed_slide_seq_HuBMAP)\n",
    "\n",
    "\n",
    "KPMP_SN_raw.obs.rename(columns=KPMP_shared, inplace=True)\n",
    "\n",
    "KPMP_var_shared = {\n",
    "    'feature_name': 'gene_name'\n",
    "}\n",
    "KPMP_SN_raw.var.rename(columns=KPMP_var_shared, inplace=True)\n",
    "\n",
    "Hubmap_shared = {\n",
    "    \"cell_id\": \"cell_id\", \n",
    "    \"predicted_CLID\": \"cl_id\",\n",
    "    \"predicted_label\": \"cl_label\",\n",
    "    \"n_genes\": \"gene_count\",\n",
    "    \"age\": \"age\",\n",
    "    \"sex\": \"sex\",\n",
    "    \"race\": \"race\",\n",
    "    \"hubmap_id\":\"dataset_id\"\n",
    "}\n",
    "HubMap_var_shared = {\n",
    "    \"hugo_symbol\": \"gene_name\"\n",
    "}\n",
    "\n",
    "def add_common_value(adata,RK):\n",
    " \n",
    "    adata.obs['as_id'] = \"UBERON:0002113\"\n",
    "    adata.obs['disease'] = \"normal\"\n",
    "    adata.obs['consortium'] = 'HuBMAP'\n",
    "    if RK == True:\n",
    "        adata.obs['collection'] = 'HuBMAP Right Kidney'\n",
    "    else:\n",
    "        adata.obs['collection'] = 'HuBMAP Left Kidney'\n",
    "    adata.obs['tissue'] = 'kidney'\n",
    "    # Rename HuBMAP columns using inplace renaming\n",
    "    adata.obs.rename(columns=Hubmap_shared, inplace=True)\n",
    "    adata.var.rename(columns=HubMap_var_shared, inplace=True)\n",
    "    adata.var.index = adata.var.index.to_series().apply(lambda x: re.sub(r'\\.\\d+$', '', x))\n",
    "    return adata\n",
    "\n",
    "HuBMAP_LK_raw = add_common_value(HuBMAP_LK_raw,False)\n",
    "HuBMAP_RK_raw = add_common_value(HuBMAP_RK_raw,True)\n",
    "\n",
    "\n",
    "def  remove_extra_var(adata):\n",
    "    columns_to_drop = [col for col in adata.var.columns if col != 'gene_name']\n",
    "    adata.var.drop(columns=columns_to_drop, inplace=True)\n",
    "    return adata\n",
    "\n",
    "KPMP_SN_raw=remove_extra_var(KPMP_SN_raw)\n",
    "\n",
    "def drop_unqiue_obs_columns(adata):\n",
    "    keep_cols = [\n",
    "    'consortium',\n",
    "    'collection',\n",
    "    'dataset_id',\n",
    "    'as_id',\n",
    "    'cl_id',\n",
    "    'cl_label',\n",
    "    # 'gene_count',\n",
    "    'age',\n",
    "    'sex',\n",
    "    'race',\n",
    "    'disease'\n",
    "    ]\n",
    "    missing_cols = [col for col in keep_cols if col not in adata.obs.columns]\n",
    "    if missing_cols:\n",
    "        print(\"Warning: The following columns are missing in adata.obs:\", missing_cols)\n",
    "    available_cols = [col for col in keep_cols if col in adata.obs.columns]\n",
    "    adata.obs = adata.obs[available_cols]\n",
    "    return adata\n",
    "\n",
    "HuBMAP_LK_raw = drop_unqiue_obs_columns(HuBMAP_LK_raw)\n",
    "HuBMAP_RK_raw = drop_unqiue_obs_columns(HuBMAP_RK_raw)\n",
    "\n",
    "# HuBMAP_combined=drop_unqiue_obs_columns(HuBMAP_combined)\n",
    "\n",
    "KPMP_SN_raw=drop_unqiue_obs_columns(KPMP_SN_raw)\n",
    "\n",
    "def map_the_gene_name(adata):\n",
    "    common_genes = KPMP_SN_raw.var.index.intersection(adata.var.index)\n",
    "    \n",
    "    gene_names = KPMP_SN_raw.var.loc[common_genes, 'gene_name'].astype(str)\n",
    "    \n",
    "    adata.var['gene_name'] = adata.var.get('gene_name', pd.Series(dtype=str)).astype(str)\n",
    "    \n",
    "    adata.var.loc[common_genes, 'gene_name'] = gene_names\n",
    "    \n",
    "    print(f\"Identified and updated {len(common_genes)} common gene IDs.\")\n",
    "    \n",
    "    \n",
    "    return adata\n",
    "\n",
    "HuBMAP_LK_raw = map_the_gene_name(HuBMAP_LK_raw)\n",
    "HuBMAP_RK_raw = map_the_gene_name(HuBMAP_RK_raw)\n",
    "\n",
    "\n",
    "shared = (\n",
    "    KPMP_SN_raw.var.index\n",
    "    .intersection(HuBMAP_LK_raw.var.index)\n",
    "    .intersection(HuBMAP_RK_raw.var.index)\n",
    ")\n",
    "\n",
    "print(f\"Found {len(shared)} genes shared across all datasets.\")\n",
    "\n",
    "KPMP_SN_raw     = KPMP_SN_raw[:, shared].copy()\n",
    "HuBMAP_LK_raw   = HuBMAP_LK_raw[:, shared].copy()\n",
    "HuBMAP_RK_raw   = HuBMAP_RK_raw[:, shared].copy()\n",
    "print(\"Obs columns before cocnat:\", KPMP_SN_raw.obs.columns.tolist())\n",
    "\n",
    "print(\"Obs columns before cocnat:\", HuBMAP_LK_raw.obs.columns.tolist())\n",
    "\n",
    "print(\"Obs columns before cocnat:\", HuBMAP_RK_raw.obs.columns.tolist())\n",
    "adata_concat_scvi = anndata.concat([KPMP_SN_raw,HuBMAP_LK_raw,HuBMAP_RK_raw], label=\"batch\", keys=[\"KPMP\", \"HuBMAP_LK\", \"HuBMAP_RK\"],axis=0, join=\"inner\", merge=\"first\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9e39bdfc-2b40-4f6a-9bba-e0e494142f8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_concat_scvi_1=adata_concat_scvi.copy()\n",
    "sc.pp.filter_cells(adata_concat_scvi, min_genes=200) \n",
    "sc.pp.filter_genes (adata_concat_scvi, min_cells=3) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c85bb8c1-3730-4503-8d84-cb941d7f8ae4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Compressed Sparse Row sparse matrix of dtype 'float32'\n",
      "\twith 1009461054 stored elements and shape (832343, 35878)>\n",
      "  Coords\tValues\n",
      "  (0, 44)\t1.0\n",
      "  (0, 74)\t1.0\n",
      "  (0, 678)\t1.0\n",
      "  (0, 1638)\t1.0\n",
      "  (0, 1949)\t1.0\n",
      "  (0, 2157)\t1.0\n",
      "  (0, 3179)\t1.0\n",
      "  (0, 3226)\t1.0\n",
      "  (0, 3594)\t1.0\n",
      "  (0, 4418)\t1.0\n",
      "  (0, 4873)\t1.0\n",
      "  (0, 5484)\t1.0\n",
      "  (0, 6447)\t1.0\n",
      "  (0, 6495)\t1.0\n",
      "  (0, 6581)\t5.0\n",
      "  (0, 6612)\t1.0\n",
      "  (0, 7172)\t1.0\n",
      "  (0, 7236)\t1.0\n",
      "  (0, 7714)\t1.0\n",
      "  (0, 7716)\t1.0\n",
      "  (0, 8081)\t1.0\n",
      "  (0, 8113)\t1.0\n",
      "  (0, 8120)\t1.0\n",
      "  (0, 8170)\t3.0\n",
      "  (0, 8206)\t1.0\n",
      "  :\t:\n",
      "  (832342, 35631)\t1.0\n",
      "  (832342, 35641)\t0.5\n",
      "  (832342, 35670)\t1.0\n",
      "  (832342, 35681)\t0.25\n",
      "  (832342, 35682)\t1.0\n",
      "  (832342, 35696)\t0.33333298563957214\n",
      "  (832342, 35714)\t0.5\n",
      "  (832342, 35741)\t1.0\n",
      "  (832342, 35755)\t0.5\n",
      "  (832342, 35782)\t1.0\n",
      "  (832342, 35785)\t1.0\n",
      "  (832342, 35786)\t0.5\n",
      "  (832342, 35796)\t0.5\n",
      "  (832342, 35808)\t0.5\n",
      "  (832342, 35809)\t0.5\n",
      "  (832342, 35811)\t1.0\n",
      "  (832342, 35812)\t1.0\n",
      "  (832342, 35813)\t1.0\n",
      "  (832342, 35830)\t0.5\n",
      "  (832342, 35835)\t0.5\n",
      "  (832342, 35839)\t0.5\n",
      "  (832342, 35840)\t0.5\n",
      "  (832342, 35848)\t1.0\n",
      "  (832342, 35850)\t0.5\n",
      "  (832342, 35875)\t1.0\n"
     ]
    }
   ],
   "source": [
    "print(adata_concat_scvi.X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "543ed495-2f2b-4585-a7fd-a00adc79f4c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata_concat_scvi,target_sum=1e4)\n",
    "# sc.pp.log1p(adata_concat_scvi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6f39178f-7249-4b10-ae7d-d82a2712e68c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                  cl_label  \\\n",
      "KC33_AAACCCAAGTAGTCAA-1  kidney interstitial alternatively activated ma...   \n",
      "KC33_AAACCCACAACTGAAA-1    kidney distal convoluted tubule epithelial cell   \n",
      "KC33_AAACCCACACGTTGGC-1                 epithelial cell of proximal tubule   \n",
      "KC33_AAACCCAGTACATACC-1                                             T cell   \n",
      "KC33_AAACCCATCCGAGCTG-1                 epithelial cell of proximal tubule   \n",
      "KC33_AAACCCATCTTCGTGC-1                 epithelial cell of proximal tubule   \n",
      "KC33_AAACGAACAACGATCT-1  kidney loop of Henle thick ascending limb epit...   \n",
      "KC33_AAACGAACACCGTGGT-1  kidney loop of Henle thick ascending limb epit...   \n",
      "KC33_AAACGAACAGTAGAAT-1                                             T cell   \n",
      "KC33_AAACGAACAGTATACC-1                 epithelial cell of proximal tubule   \n",
      "KC33_AAACGAAGTAGGCAGT-1                 epithelial cell of proximal tubule   \n",
      "KC33_AAACGAAGTGCATGAG-1                 epithelial cell of proximal tubule   \n",
      "KC33_AAACGCTAGCTAGTTC-1           kidney collecting duct intercalated cell   \n",
      "KC33_AAACGCTCATAACTCG-1           kidney collecting duct intercalated cell   \n",
      "KC33_AAACGCTCATATTCGG-1                 epithelial cell of proximal tubule   \n",
      "KC33_AAACGCTGTACGCTAT-1                 epithelial cell of proximal tubule   \n",
      "KC33_AAACGCTGTCGTATGT-1  kidney loop of Henle thick ascending limb epit...   \n",
      "KC33_AAACGCTGTTTCGCTC-1                     kidney interstitial fibroblast   \n",
      "KC33_AAACGCTTCAAGCCTA-1                 epithelial cell of proximal tubule   \n",
      "KC33_AAACGCTTCCTATTGT-1                 epithelial cell of proximal tubule   \n",
      "\n",
      "                                                          general_cl_label  \n",
      "KC33_AAACCCAAGTAGTCAA-1  Kidney interstitial alternatively activated ma...  \n",
      "KC33_AAACCCACAACTGAAA-1           Distal convoluted tubule epithelial cell  \n",
      "KC33_AAACCCACACGTTGGC-1                    Proximal tubule epithelial cell  \n",
      "KC33_AAACCCAGTACATACC-1                                             T cell  \n",
      "KC33_AAACCCATCCGAGCTG-1                    Proximal tubule epithelial cell  \n",
      "KC33_AAACCCATCTTCGTGC-1                    Proximal tubule epithelial cell  \n",
      "KC33_AAACGAACAACGATCT-1  kidney loop of Henle thick ascending limb epit...  \n",
      "KC33_AAACGAACACCGTGGT-1  kidney loop of Henle thick ascending limb epit...  \n",
      "KC33_AAACGAACAGTAGAAT-1                                             T cell  \n",
      "KC33_AAACGAACAGTATACC-1                    Proximal tubule epithelial cell  \n",
      "KC33_AAACGAAGTAGGCAGT-1                    Proximal tubule epithelial cell  \n",
      "KC33_AAACGAAGTGCATGAG-1                    Proximal tubule epithelial cell  \n",
      "KC33_AAACGCTAGCTAGTTC-1                  Collecting duct intercalated cell  \n",
      "KC33_AAACGCTCATAACTCG-1                  Collecting duct intercalated cell  \n",
      "KC33_AAACGCTCATATTCGG-1                    Proximal tubule epithelial cell  \n",
      "KC33_AAACGCTGTACGCTAT-1                    Proximal tubule epithelial cell  \n",
      "KC33_AAACGCTGTCGTATGT-1  kidney loop of Henle thick ascending limb epit...  \n",
      "KC33_AAACGCTGTTTCGCTC-1                     Kidney interstitial fibroblast  \n",
      "KC33_AAACGCTTCAAGCCTA-1                    Proximal tubule epithelial cell  \n",
      "KC33_AAACGCTTCCTATTGT-1                    Proximal tubule epithelial cell  \n"
     ]
    }
   ],
   "source": [
    "label_mapping = {\n",
    "    # Proximal tubule\n",
    "    \"epithelial cell of proximal tubule\": \"Proximal tubule epithelial cell\",\n",
    "    \"epithelial cell of proximal tubule segment 1\": \"Proximal tubule epithelial cell\",\n",
    "    \"epithelial cell of proximal tubule segment 2\": \"Proximal tubule epithelial cell\",\n",
    "    \"epithelial cell of proximal tubule segment 3\": \"Proximal tubule epithelial cell\",\n",
    "\n",
    "    # Distal convoluted tubule\n",
    "    \"kidney distal convoluted tubule epithelial cell\": \"Distal convoluted tubule epithelial cell\",\n",
    "    \"epithelial cell of early distal convoluted tubule\": \"Distal convoluted tubule epithelial cell\",\n",
    "    \"epithelial cell of late distal convoluted tubule\": \"Distal convoluted tubule epithelial cell\",\n",
    "\n",
    "    # Loop of Henle\n",
    "    \"kidney loop of Henle thin descending limb epithelial cell\": \"Loop of Henle epithelial cell \",\n",
    "    \"kidney loop of Henle thin ascending limb epithelial cell\": \"Loop of Henle epithelial cell \",\n",
    "    \"kidney loop of Henle cortical thick ascending limb epithelial cell\": \"Loop of Henle epithelial cell \",\n",
    "    \"kidney loop of Henle medullary thick ascending limb epithelial cell\": \"Loop of Henle epithelial cell \",\n",
    "    \"kidney loop of Henle long descending thin limb outer medulla epithelial cell\": \"Loop of Henle epithelial cell \",\n",
    "    \"kidney loop of Henle long descending thin limb inner medulla epithelial cell\": \"Loop of Henle epithelial cell \",\n",
    "    \"kidney loop of Henle short descending thin limb epithelial cell\": \"Loop of Henle epithelial cell \",\n",
    "\n",
    "    # Collecting duct principal cells\n",
    "    \"kidney collecting duct principal cell\": \"Collecting duct principal cell\",\n",
    "    \"kidney cortex collecting duct principal cell\": \"Collecting duct principal cell\",\n",
    "    \"kidney outer medulla collecting duct principal cell\": \"Collecting duct principal cell\",\n",
    "    \"kidney inner medulla collecting duct principal cell\": \"Collecting duct principal cell\",\n",
    "\n",
    "    # Collecting duct intercalated cells\n",
    "    \"kidney collecting duct intercalated cell\": \"Collecting duct intercalated cell\",\n",
    "    \"kidney cortex collecting duct intercalated cell\": \"Collecting duct intercalated cell\",\n",
    "    \"kidney outer medulla collecting duct intercalated cell\": \"Collecting duct intercalated cell\",\n",
    "\n",
    "    # Connecting tubule\n",
    "    \"kidney connecting tubule epithelial cell\": \"Connecting tubule epithelial cell\",\n",
    "    \"kidney connecting tubule principal cell\": \"Connecting tubule epithelial cell\",\n",
    "    \"kidney connecting tubule alpha-intercalated cell\": \"Connecting tubule epithelial cell\",\n",
    "\n",
    "    # Other epithelial\n",
    "    \"podocyte\": \"Podocyte\",\n",
    "    \"parietal epithelial cell\": \"Parietal epithelial cell\",\n",
    "    \"papillary tips cell\": \"Papillary tips cell\",\n",
    "\n",
    "    # Stromal & endothelial\n",
    "    \"endothelial cell\": \"Endothelial cell\",\n",
    "    \"kidney interstitial cell\": \"Kidney interstitial cell\",\n",
    "    \"kidney interstitial fibroblast\": \"Kidney interstitial fibroblast\",\n",
    "    \"kidney interstitial alternatively activated macrophage\": \"Kidney interstitial alternatively activated macrophage\",\n",
    "    \"renal interstitial pericyte\": \"Renal interstitial pericyte\",\n",
    "\n",
    "    # Immune cells\n",
    "    \"T cell\": \"T cell\",\n",
    "    \"cytotoxic T cell\": \"T cell\",\n",
    "    \"mature NK T cell\": \"T cell\",\n",
    "    \"B cell\": \"B cell\",\n",
    "    \"plasma cell\": \"Plasma cell\",\n",
    "    \"natural killer cell\": \"Natural killer cell\",\n",
    "    \"monocyte\": \"Monocyte\",\n",
    "    \"non-classical monocyte\": \"Monocyte\",\n",
    "    \"conventional dendritic cell\": \"Dendritic cell\",\n",
    "    \"plasmacytoid dendritic cell, human\": \"Dendritic cell\",\n",
    "    \"mononuclear phagocyte\": \"Mononuclear phagocyte\",\n",
    "    \"mast cell\": \"Mast cell\",\n",
    "    \"neutrophil\": \"Neutrophil\",\n",
    "}\n",
    "\n",
    "adata_concat_scvi.obs[\"general_cl_label\"] = adata_concat_scvi.obs[\"cl_label\"].map(label_mapping)\n",
    "\n",
    "adata_concat_scvi.obs[\"general_cl_label\"] = adata_concat_scvi.obs[\"general_cl_label\"].fillna(adata_concat_scvi.obs[\"cl_label\"])\n",
    "adata_concat_scvi.obs['age'] = adata_concat_scvi.obs['age'].astype(str)\n",
    "\n",
    "print(adata_concat_scvi.obs[[\"cl_label\", \"general_cl_label\"]].head(20))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "982542c4-e9cf-49f4-b2a1-4ee2484b3f7e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Kidney interstitial alternatively activated macrophage',\n",
       "       'Distal convoluted tubule epithelial cell',\n",
       "       'Proximal tubule epithelial cell', 'T cell',\n",
       "       'kidney loop of Henle thick ascending limb epithelial cell',\n",
       "       'Collecting duct intercalated cell',\n",
       "       'Kidney interstitial fibroblast',\n",
       "       'Connecting tubule epithelial cell',\n",
       "       'Collecting duct principal cell', 'Endothelial cell', 'B cell',\n",
       "       'Mononuclear phagocyte', 'Podocyte', 'Parietal epithelial cell',\n",
       "       'Mast cell', 'Dendritic cell', 'Plasma cell', 'Monocyte',\n",
       "       'Loop of Henle epithelial cell ', 'Renal interstitial pericyte',\n",
       "       'Neutrophil', 'Papillary tips cell', 'neural cell'], dtype=object)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_concat_scvi.obs[\"general_cl_label\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6b33c353-641c-4544-b2d3-cc2a19e2caae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Cells: 832,343\n",
      "Total Genes: 35,878\n"
     ]
    }
   ],
   "source": [
    "print(f\"Total Cells: {adata_concat_scvi.n_obs:,}\")\n",
    "print(f\"Total Genes: {adata_concat_scvi.n_vars:,}\")  \n",
    "adata_concat_scvi.write_h5ad(\"/scratch/yongqunh_root/yongqunh0/wruopeng/HuBMAP_KPMP/kpmp_hubmap_combined_after_1e4_normalization_only_afterQC.h5ad\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "41a5bd00-1407-45f1-9caf-5a359b6a751a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Cells: 832,343\n",
      "Total Genes: 35,878\n"
     ]
    }
   ],
   "source": [
    "print(f\"Total Cells: {adata_concat_scvi.n_obs:,}\")\n",
    "print(f\"Total Genes: {adata_concat_scvi.n_vars:,}\")  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7848aa07-25cf-459a-ae11-783e2282f473",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:845: UserWarning: \n",
      "AnnData expects .var.index to contain strings, but got values like:\n",
      "    ['A1BG', 'A1BG-AS1', 'A1CF', 'A2M', 'A2M-AS1']\n",
      "\n",
      "    Inferred to be: categorical\n",
      "\n",
      "  names = self._prep_dim_index(names, \"var\")\n"
     ]
    }
   ],
   "source": [
    "adata_concat_scvi.var_names=adata_concat_scvi.var['gene_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2931372c-010a-471b-80bf-802ca45630ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved as combined_clean.csv\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "csv1 = \"uberonID.csv\"\n",
    "csv2 = \"cl_mapping_edge.csv\"\n",
    "csv3 = \"genes_nodes.csv\"\n",
    "\n",
    "def load_and_fix(path):\n",
    "    df = pd.read_csv(path)\n",
    "    df.columns = [c.strip().lower() for c in df.columns]\n",
    "    df = df.rename(columns={\"types\": \"type\"})\n",
    "    return df\n",
    "\n",
    "df1 = load_and_fix(csv1)\n",
    "df2 = load_and_fix(csv2)\n",
    "df3 = load_and_fix(csv3)\n",
    "\n",
    "common_cols = set(df1.columns) & set(df2.columns) & set(df3.columns)\n",
    "df1 = df1[list(common_cols)]\n",
    "df2 = df2[list(common_cols)]\n",
    "df3 = df3[list(common_cols)]\n",
    "\n",
    "combined = pd.concat([df1, df2, df3], ignore_index=True).drop_duplicates()\n",
    "\n",
    "# Save\n",
    "combined.to_csv(\"combined_nodes.csv\", index=False)\n",
    "\n",
    "print(\"Saved as combined_clean.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d937a2b3-b19d-4170-a98d-75df88727f78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved uberon_mapping.csv with 23 rows (UBERON only + kidney fallback).\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import requests\n",
    "import pandas as pd\n",
    "\n",
    "FIXED_TYPE   = \"http://purl.obolibrary.org/obo/UBERON_0000061\"\n",
    "FIXED_SOURCE = \"https://purl.humanatlas.io/vocab/hp\"\n",
    "\n",
    "_SUFFIXES = [\n",
    "    r\"\\bepithelial cells?\\b\", r\"\\bepithelial cell\\b\",\n",
    "    r\"\\bprincipal cells?\\b\",  r\"\\bintercalated cells?\\b\",\n",
    "    r\"\\bfibroblasts?\\b\",      r\"\\bpericytes?\\b\",\n",
    "    r\"\\bendothelial cells?\\b\",r\"\\bmesangial cells?\\b\",\n",
    "    r\"\\bpodocytes?\\b\",        r\"\\bimmune cells?\\b\",\n",
    "    r\"\\b(neutrophils?|monocytes?|macrophages?|dendritic cells?)\\b\",\n",
    "    r\"\\b(T|B) cells?\\b\",      r\"\\bstem cells?\\b\",\n",
    "    r\"\\bplasma cells?\\b\",     r\"\\bmast cells?\\b\",\n",
    "    r\"\\bneurons?\\b\",          r\"\\bglial cells?\\b\",\n",
    "    r\"\\binterstitial cells?\\b\", r\"\\bmyofibroblasts?\\b\",\n",
    "    r\"\\bprincipal cell\\b\",    r\"\\bintercalated cell\\b\",\n",
    "    r\"\\bcell(?:s)?\\b\"\n",
    "]\n",
    "_SUFFIX_RE = re.compile(\"|\".join(_SUFFIXES), flags=re.IGNORECASE)\n",
    "\n",
    "def _clean_term_for_uberon(term: str) -> str:\n",
    "    cleaned = _SUFFIX_RE.sub(\"\", term)\n",
    "    cleaned = re.sub(r\"\\s+\", \" \", cleaned).strip(\" ,;.-\").strip()\n",
    "    return cleaned or term  # never return empty\n",
    "\n",
    "def _search_ols_uberon(q: str):\n",
    "    \"\"\"Call OLS search and return docs list (or []).\"\"\"\n",
    "    url = \"https://www.ebi.ac.uk/ols/api/search\"\n",
    "    params = {\n",
    "        \"q\": q,\n",
    "        \"ontology\": \"uberon\",  \n",
    "        \"type\": \"class\",       \n",
    "        \"exact\": \"false\",\n",
    "        \"rows\": 50\n",
    "    }\n",
    "    try:\n",
    "        r = requests.get(url, params=params, timeout=20)\n",
    "        if r.status_code == 200:\n",
    "            j = r.json()\n",
    "            return j.get(\"response\", {}).get(\"docs\", []) or []\n",
    "    except Exception:\n",
    "        pass\n",
    "    return []\n",
    "\n",
    "def _pick_uberon_hit(docs):\n",
    "    \"\"\"Return the first doc that is definitely UBERON.\"\"\"\n",
    "    for d in docs:\n",
    "        iri = d.get(\"iri\", \"\") or \"\"\n",
    "        obo = d.get(\"obo_id\", \"\") or \"\"\n",
    "        onto = (d.get(\"ontology_name\") or \"\").lower()\n",
    "        if onto == \"uberon\" or \"UBERON_\" in iri or obo.startswith(\"UBERON:\"):\n",
    "            return d\n",
    "    return None\n",
    "\n",
    "def get_uberon_entry_strict(term: str):\n",
    "    \"\"\"\n",
    "    Return dict with iri, label, type, source.\n",
    "    Try original term, then a 'cleaned' term.\n",
    "    If no UBERON hit is found, return kidney default.\n",
    "    \"\"\"\n",
    "    term = \"\" if term is None else str(term).strip()\n",
    "    # 1) try raw\n",
    "    docs = _search_ols_uberon(term)\n",
    "    hit = _pick_uberon_hit(docs)\n",
    "\n",
    "    if hit is None:\n",
    "        cleaned = _clean_term_for_uberon(term)\n",
    "        if cleaned and cleaned.lower() != term.lower():\n",
    "            docs2 = _search_ols_uberon(cleaned)\n",
    "            hit = _pick_uberon_hit(docs2)\n",
    "\n",
    "    if hit:\n",
    "        return {\n",
    "            \"iri\": hit.get(\"iri\"),\n",
    "            \"label\": hit.get(\"label\"),\n",
    "            \"type\": FIXED_TYPE,\n",
    "            \"source\": FIXED_SOURCE,\n",
    "        }\n",
    "\n",
    "    return {\n",
    "        \"iri\": \"http://purl.obolibrary.org/obo/UBERON_0002113\",\n",
    "        \"label\": \"kidney\",\n",
    "        \"type\": FIXED_TYPE,\n",
    "        \"source\": FIXED_SOURCE,\n",
    "    }\n",
    "\n",
    "terms = pd.unique(adata_concat_scvi.obs['general_cl_label']).tolist()\n",
    "\n",
    "rows = [get_uberon_entry_strict(t) for t in terms]\n",
    "df = pd.DataFrame(rows, columns=[\"iri\", \"label\", \"type\", \"source\"])\n",
    "df.to_csv(\"uberon_mapping_edge1.csv\", index=False)\n",
    "print(f\"✅ Saved uberon_mapping.csv with {len(df)} rows (UBERON only + kidney fallback).\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6de9db1c-7546-415b-bdb1-41395c09142c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes.csv created with real HGNC iris.\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "\n",
    "def get_hgnc_iri(gene_symbol: str) -> str:\n",
    "    \"\"\"\n",
    "    Query Identifiers.org / HGNC to resolve a gene symbol to its iri.\n",
    "    Falls back to empty string if not found.\n",
    "    \"\"\"\n",
    "    url = f\"https://rest.genenames.org/fetch/symbol/{gene_symbol}\"\n",
    "    headers = {\"Accept\": \"application/json\"}\n",
    "    \n",
    "    try:\n",
    "        r = requests.get(url, headers=headers, timeout=10)\n",
    "        r.raise_for_status()\n",
    "        docs = r.json().get(\"response\", {}).get(\"docs\", [])\n",
    "        if docs:\n",
    "            hgnc_id = docs[0].get(\"hgnc_id\", \"\")\n",
    "            if hgnc_id:\n",
    "                # hgnc_id is like \"HGNC:132\", so we split and take the number\n",
    "                number = hgnc_id.split(\":\")[1]\n",
    "                return f\"http://identifiers.org/hgnc/{number}\"\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to fetch {gene_symbol}: {e}\")\n",
    "    return \"\"\n",
    "\n",
    "# gene_names = adata_concat_scvi.var['gene_name'].tolist()\n",
    "gene_names = [        \"CCL21\",\n",
    "    \"NRXN1\",\n",
    "    \"SLC9A3\",\n",
    "    \"SPP1\",\n",
    "    \"SERPINE2\",\n",
    "    \"IL18\",\n",
    "    \"NBAS\",\n",
    "    \"TIMP2\",\n",
    "    \"IGFBP7\",\n",
    "    \"SUGP1\",\n",
    "    \"APOA1\",\n",
    "    \"MMP9\",\n",
    "    \"ITGAM\",\n",
    "    \"TFF3\",\n",
    "    \"B2M\",\n",
    "    \"S100A6\",\n",
    "    \"IL10\",\n",
    "    \"IL6\",\n",
    "    \"CXCL8\",\n",
    "    \"LCN2\",\n",
    "    \"HAVCR1\",\n",
    "    \"CST3\",\n",
    "    \"CCL2\",\n",
    "    \"FABP1\"]\n",
    "\n",
    "\n",
    "iris = [get_hgnc_iri(g) for g in gene_names]\n",
    "\n",
    "df = pd.DataFrame({\n",
    "    \"iri\": iris,\n",
    "    \"label\": gene_names,\n",
    "    \"type\": [\"http://purl.bioontology.org/ontology/HGNC/gene\"] * len(gene_names),\n",
    "    \"source\": [\"https://purl.humanatlas.io/vocab/hp\"] * len(gene_names)\n",
    "})\n",
    "\n",
    "df.to_csv(\"genes_nodes.csv\", index=False)\n",
    "print(\"genes.csv created with real HGNC iris.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d2c9320-d271-4e66-8438-6e03e22940d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved cl_mapping.csv with 23 unique cell type mappings\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def get_cl_info(term):\n",
    "    \"\"\"Query OLS for a Cell Ontology (CL) term given a cell type name.\"\"\"\n",
    "    url = \"https://www.ebi.ac.uk/ols/api/search\"\n",
    "    params = {\"q\": term, \"ontology\": \"cl\", \"exact\": \"false\"}\n",
    "    r = requests.get(url, params=params)\n",
    "\n",
    "    if r.status_code == 200:\n",
    "        data = r.json()\n",
    "        if data['response']['numFound'] > 0:\n",
    "            hit = data['response']['docs'][0]\n",
    "            return {\n",
    "                \"iri\": hit.get(\"iri\"),\n",
    "                \"label\": hit.get(\"label\"),\n",
    "                \"types\": \"http://purl.obolibrary.org/obo/CL_0000000\",  # e.g. CL_0000236\n",
    "                \"source\": \"https://www.ebi.ac.uk/ols/ontologies/cl\"\n",
    "            }\n",
    "    return {\n",
    "        \"iri\": \"NA\",\n",
    "        \"label\": \"NA\",\n",
    "        \"types\": \"http://purl.obolibrary.org/obo/CL_0000000\",\n",
    "        \"source\": \"https://www.ebi.ac.uk/ols/ontologies/cl\"\n",
    "    }\n",
    "\n",
    "terms = adata_concat_scvi.obs['general_cl_label'].unique().tolist()\n",
    "\n",
    "results = [get_cl_info(t) for t in terms]\n",
    "\n",
    "df = pd.DataFrame(results)\n",
    "\n",
    "df.to_csv(\"cl_mapping_edge.csv\", index=False)\n",
    "\n",
    "print(\"✅ Saved cl_mapping.csv with\", len(df), \"unique cell type mappings\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7b6c23b4-7961-4631-b610-ea5c06fadddb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AKI subset: (41512, 35878)\n",
      "CKD subset: (62983, 35878)\n",
      "Normal subset: (58754, 35878)\n"
     ]
    }
   ],
   "source": [
    "adata_fem = adata_concat_scvi[adata_concat_scvi.obs[\"sex\"] == \"male\"].copy()\n",
    "\n",
    "adata_aki = adata_fem[adata_fem.obs[\"disease\"] == \"acute kidney failure\"].copy()\n",
    "adata_ckd = adata_fem[adata_fem.obs[\"disease\"] == \"chronic kidney disease\"].copy()\n",
    "adata_normal = adata_fem[adata_fem.obs[\"disease\"] == \"normal\"].copy()\n",
    "\n",
    "# Print sizes\n",
    "print(\"AKI subset:\", adata_aki.shape)      \n",
    "print(\"CKD subset:\", adata_ckd.shape)\n",
    "print(\"Normal subset:\", adata_normal.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a4fe0bff-da86-44dd-a8f9-6c8422401b70",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['KC33', 'KC34', 'KC35', 'KC36', 'PA1', ..., 'HBM445.HBRQ.488', 'HBM729.TVMN.534', 'HBM478.VWXX.362', 'HBM847.MDSJ.826', 'HBM766.NZWP.682']\n",
      "Length: 181\n",
      "Categories (181, object): ['HBM226.HRTZ.874', 'HBM247.HLXR.494', 'HBM253.ZBGF.863', 'HBM264.MJCH.639', ..., 'PA1', 'PA3', 'PA4', 'PA5']\n"
     ]
    }
   ],
   "source": [
    "print(adata_concat_scvi.obs['dataset_id'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1861188a-d8b0-4c15-a7a6-f80ab2a7475c",
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_genes=[        \"CCL21\",\n",
    "    \"NRXN1\",\n",
    "    \"SLC9A3\",\n",
    "    \"SPP1\",\n",
    "    \"SERPINE2\",\n",
    "    \"IL18\",\n",
    "    \"NBAS\",\n",
    "    \"TIMP2\",\n",
    "    \"IGFBP7\",\n",
    "    \"SUGP1\",\n",
    "    \"APOA1\",\n",
    "    \"MMP9\",\n",
    "    \"ITGAM\",\n",
    "    \"TFF3\",\n",
    "    \"B2M\",\n",
    "    \"S100A6\",\n",
    "    \"IL10\",\n",
    "    \"IL6\",\n",
    "    \"CXCL8\",\n",
    "    \"LCN2\",\n",
    "    \"HAVCR1\",\n",
    "    \"CST3\",\n",
    "    \"CCL2\",\n",
    "    \"FABP1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "9ee67112-ff8a-4be3-bdda-649acae610a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/wruopeng/.local/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:459: RuntimeWarning: overflow encountered in expm1\n",
      "  self.expm1_func(mean_rest) + 1e-9\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:461: RuntimeWarning: divide by zero encountered in log2\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:459: RuntimeWarning: overflow encountered in expm1\n",
      "  self.expm1_func(mean_rest) + 1e-9\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:461: RuntimeWarning: divide by zero encountered in log2\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n"
     ]
    }
   ],
   "source": [
    "sc.tl.rank_genes_groups(\n",
    "    adata_concat_scvi,\n",
    "    groupby=\"disease\",        \n",
    "    method=\"t-test\",          \n",
    "    reference=\"normal\"           \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "7978d041-4c61-4ceb-b466-e4f216101c19",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "unterminated string literal (detected at line 1) (951358321.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[88], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    header=['biomarker_id','biomarker_label','cell_types','disease_types','mean_expression_value\",\"fold changes\"]\u001b[0m\n\u001b[0m                                                                          ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m unterminated string literal (detected at line 1)\n"
     ]
    }
   ],
   "source": [
    "header=['biomarker_id','biomarker_label','cell_types','disease_types','mean_expression_value\",\"fold changes\"]\n",
    "df_KPMP_HuBMAP_integrated = pd.DataFrame(columns=header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6a49063c-ca1d-4b98-8bb1-c4589cbf644a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['normal', 'chronic kidney disease', 'acute kidney failure']\n",
      "Categories (3, object): ['acute kidney failure', 'chronic kidney disease', 'normal']\n"
     ]
    }
   ],
   "source": [
    "print(adata_concat_scvi.obs['disease'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59badf4f-d9eb-42bd-ab43-e77dd2b1f30c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                             iri  \\\n",
      "0  http://purl.obolibrary.org/obo/UBERON_0005215   \n",
      "1  http://purl.obolibrary.org/obo/UBERON_0001292   \n",
      "2  http://purl.obolibrary.org/obo/UBERON_0004134   \n",
      "3  http://purl.obolibrary.org/obo/UBERON_0010393   \n",
      "4  http://purl.obolibrary.org/obo/UBERON_0001291   \n",
      "\n",
      "                                               label  \\\n",
      "0                                             kidney   \n",
      "1         kidney distal convoluted tubule epithelial   \n",
      "2                  kidney proximal tubule epithelial   \n",
      "3                                             kidney   \n",
      "4  kidney kidney loop of henle thick ascending li...   \n",
      "\n",
      "                                            type  \\\n",
      "0  http://purl.obolibrary.org/obo/UBERON_0000061   \n",
      "1  http://purl.obolibrary.org/obo/UBERON_0000061   \n",
      "2  http://purl.obolibrary.org/obo/UBERON_0000061   \n",
      "3  http://purl.obolibrary.org/obo/UBERON_0000061   \n",
      "4  http://purl.obolibrary.org/obo/UBERON_0000061   \n",
      "\n",
      "                                source  \\\n",
      "0  https://purl.humanatlas.io/vocab/hp   \n",
      "1  https://purl.humanatlas.io/vocab/hp   \n",
      "2  https://purl.humanatlas.io/vocab/hp   \n",
      "3  https://purl.humanatlas.io/vocab/hp   \n",
      "4  https://purl.humanatlas.io/vocab/hp   \n",
      "\n",
      "                                           cell_type  \\\n",
      "0  Kidney interstitial alternatively activated ma...   \n",
      "1           Distal convoluted tubule epithelial cell   \n",
      "2                    Proximal tubule epithelial cell   \n",
      "3                                             T cell   \n",
      "4  kidney loop of Henle thick ascending limb epit...   \n",
      "\n",
      "                                              tissue  \n",
      "0  Kidney interstitial alternatively activated ma...  \n",
      "1                Distal convoluted tubule epithelial  \n",
      "2                         Proximal tubule epithelial  \n",
      "3                                                  T  \n",
      "4  kidney loop of Henle thick ascending limb epit...  \n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "import requests\n",
    "import pandas as pd\n",
    "\n",
    "def get_uberon_info(term):\n",
    "    \"\"\"Query OLS for a UBERON term given a tissue/structure name.\"\"\"\n",
    "    url = \"https://www.ebi.ac.uk/ols/api/search\"\n",
    "    params = {\"q\": term, \"ontology\": \"uberon\", \"exact\": \"false\"}\n",
    "    r = requests.get(url, params=params)\n",
    "\n",
    "    if r.status_code == 200:\n",
    "        data = r.json()\n",
    "        if data['response']['numFound'] > 0:\n",
    "            hit = data['response']['docs'][0]\n",
    "            return {\n",
    "                \"iri\": hit.get(\"iri\"),\n",
    "                \"label\": hit.get(\"label\"),\n",
    "                \"type\": \"http://purl.obolibrary.org/obo/UBERON_0000061\",\n",
    "                \"source\": \"https://purl.humanatlas.io/vocab/hp\"\n",
    "            }\n",
    "    return {\n",
    "        \"iri\": \"http://purl.obolibrary.org/obo/UBERON_0002113\",\n",
    "        \"label\": \"kidney\",\n",
    "        \"type\": \"UBERON:0002113\",\n",
    "        \"source\": \"https://purl.humanatlas.io/vocab/hp\"\n",
    "    }\n",
    "\n",
    "def map_celltypes_to_uberon(df_in):\n",
    "    results = []\n",
    "    for cl_label, tissue in zip(df_in[\"cl_label\"], df_in[\"tissue\"]):\n",
    "        \n",
    "        info = get_uberon_info(tissue)\n",
    "        if any(x in cl_label.lower() for x in [\n",
    "            \"proximal\", \"distal\", \"loop of henle\", \"collecting\", \n",
    "            \"podocyte\", \"parietal\", \"fibroblast\", \"pericyte\", \"papillary\", \"interstitium\"\n",
    "        ]):\n",
    "            label_val = \"kidney \" + tissue.lower()\n",
    "        else:\n",
    "            label_val = \"kidney\"\n",
    "\n",
    "        info[\"cell_type\"] = cl_label\n",
    "        info[\"tissue\"] = tissue\n",
    "        info[\"label\"] = label_val  \n",
    "        results.append(info)\n",
    "    return pd.DataFrame(results)\n",
    "\n",
    "celltypes = list(adata_concat_scvi.obs['general_cl_label'].unique())\n",
    "df_input = pd.DataFrame({\n",
    "    \"cl_label\": celltypes,\n",
    "    \"tissue\": [name.replace(\"cell\", \"\").replace(\"Cell\", \"\").strip() for name in celltypes]\n",
    "})\n",
    "\n",
    "df_UBERON = map_celltypes_to_uberon(df_input)\n",
    "\n",
    "df_UBERON.to_csv(\"celltypes_uberon.csv\", index=False, quoting=1)\n",
    "print(df_UBERON.head())\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6fcf559-e8f4-41e4-b642-b0818832b25b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def mean_expression_compare_to_normal(adata, celltype_key, disease_key,\n",
    "                                      disease_list, selected_celltypes=None, selected_genes=None,\n",
    "                                      save_path=None):\n",
    "    results = []\n",
    "\n",
    "    for disease in disease_list:\n",
    "        disease_mask = (adata.obs[disease_key] == disease)\n",
    "        if selected_celltypes is not None:\n",
    "            disease_mask &= adata.obs[celltype_key].isin(selected_celltypes)\n",
    "        adata_dis = adata[disease_mask, :]\n",
    "\n",
    "        normal_mask = (adata.obs[disease_key] == \"normal\")\n",
    "        if selected_celltypes is not None:\n",
    "            normal_mask &= adata.obs[celltype_key].isin(selected_celltypes)\n",
    "        adata_norm = adata[normal_mask, :]\n",
    "\n",
    "        if selected_genes is not None:\n",
    "            adata_dis = adata_dis[:, selected_genes]\n",
    "            adata_norm = adata_norm[:, selected_genes]\n",
    "\n",
    "        for celltype in adata_dis.obs[celltype_key].unique():\n",
    "            idx_dis = np.where(adata_dis.obs[celltype_key] == celltype)[0]\n",
    "            idx_norm = np.where(adata_norm.obs[celltype_key] == celltype)[0]\n",
    "\n",
    "            if len(idx_dis) == 0 or len(idx_norm) == 0:\n",
    "                continue\n",
    "\n",
    "            X_dis = adata_dis[idx_dis, :].X\n",
    "            X_norm = adata_norm[idx_norm, :].X\n",
    "\n",
    "            mean_dis = np.array(X_dis.mean(axis=0)).ravel()\n",
    "            mean_norm = np.array(X_norm.mean(axis=0)).ravel()\n",
    "            \n",
    "            dis_sex_counts = adata_dis.obs.iloc[idx_dis][\"sex\"].value_counts().to_dict()\n",
    "            norm_sex_counts = adata_norm.obs.iloc[idx_norm][\"sex\"].value_counts().to_dict()\n",
    "\n",
    "            num_female_dis = dis_sex_counts.get(\"female\", 0)\n",
    "            num_male_dis = dis_sex_counts.get(\"male\", 0)\n",
    "            num_female_norm = norm_sex_counts.get(\"female\", 0)\n",
    "            num_male_norm = norm_sex_counts.get(\"male\", 0)\n",
    "            \n",
    "            dis_datasets = adata_dis.obs.iloc[idx_dis]['dataset_id'].unique().tolist()\n",
    "            norm_datasets = adata_norm.obs.iloc[idx_norm]['dataset_id'].unique().tolist()\n",
    "\n",
    "            num_dis_datasets = len(dis_datasets)\n",
    "            num_norm_datasets = len(norm_datasets)\n",
    "\n",
    "            for gene, val_dis, val_norm in zip(adata_dis.var_names, mean_dis, mean_norm):\n",
    "                results.append({\n",
    "                    \"gene\": gene,\n",
    "                    \"disease\": disease,\n",
    "                    \"cell_type\": celltype,\n",
    "                    \"as_id\": as_id,\n",
    "                    \"mean_expression_disease\": val_dis,\n",
    "                    \"mean_expression_normal\": val_norm,\n",
    "                    \"disease_sample_size\": len(idx_dis),\n",
    "                    \"normal_sample_size\": len(idx_norm),\n",
    "                    \"disease_female_sample_size\": num_female_dis,\n",
    "                    \"disease_male_sample_size\": num_male_dis,\n",
    "                    \"normal_female_sample_size\": num_female_norm,\n",
    "                    \"normal_male_sample_size\": num_male_norm,\n",
    "                    \"disease_num_datasets\": num_dis_datasets,\n",
    "                    \"normal_num_datasets\": num_norm_datasets,\n",
    "                    \"disease_datasets\": \",\".join(dis_datasets),\n",
    "                    \"normal_datasets\": \",\".join(norm_datasets),\n",
    "                })\n",
    "\n",
    "\n",
    "    tidy_df = pd.DataFrame(results)\n",
    "\n",
    "    if save_path:\n",
    "        tidy_df.to_csv(save_path, index=False)\n",
    "        print(f\"✅ Saved to {save_path}\")\n",
    "\n",
    "    return tidy_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "4c5355be-7af8-4032-a569-7ebd0b5db0c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['consortium', 'collection', 'dataset_id', 'as_id', 'cl_id', 'cl_label',\n",
      "       'age', 'sex', 'race', 'disease', 'batch', 'n_genes',\n",
      "       'general_cl_label'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(adata_concat_scvi.obs.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "bc98ba6a-c4b5-4f78-935f-4f2e38fb0ca6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Kidney interstitial alternatively activated macrophage'\n",
      " 'Distal convoluted tubule epithelial cell'\n",
      " 'Proximal tubule epithelial cell' 'T cell'\n",
      " 'kidney loop of Henle thick ascending limb epithelial cell'\n",
      " 'Collecting duct intercalated cell' 'Kidney interstitial fibroblast'\n",
      " 'Connecting tubule epithelial cell' 'Collecting duct principal cell'\n",
      " 'Endothelial cell' 'B cell' 'Mononuclear phagocyte' 'Podocyte'\n",
      " 'Parietal epithelial cell' 'Mast cell' 'Dendritic cell' 'Plasma cell'\n",
      " 'Monocyte' 'Loop of Henle epithelial cell ' 'Renal interstitial pericyte'\n",
      " 'Neutrophil' 'Papillary tips cell' 'neural cell']\n",
      "23\n"
     ]
    }
   ],
   "source": [
    "print(adata_concat_scvi.obs['general_cl_label'].unique())\n",
    "print(len(adata_concat_scvi.obs['general_cl_label'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "5e0a89bb-2eb7-4ce3-8b85-17f0d3345e62",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved to disease_vs_normal_mean_expression_long.csv\n",
      "          gene                 disease                        cell_type  \\\n",
      "0        CCL21  chronic kidney disease  Proximal tubule epithelial cell   \n",
      "1        NRXN1  chronic kidney disease  Proximal tubule epithelial cell   \n",
      "2       SLC9A3  chronic kidney disease  Proximal tubule epithelial cell   \n",
      "3         SPP1  chronic kidney disease  Proximal tubule epithelial cell   \n",
      "4     SERPINE2  chronic kidney disease  Proximal tubule epithelial cell   \n",
      "...        ...                     ...                              ...   \n",
      "1145      LCN2    acute kidney failure              Papillary tips cell   \n",
      "1146    HAVCR1    acute kidney failure              Papillary tips cell   \n",
      "1147      CST3    acute kidney failure              Papillary tips cell   \n",
      "1148      CCL2    acute kidney failure              Papillary tips cell   \n",
      "1149     FABP1    acute kidney failure              Papillary tips cell   \n",
      "\n",
      "      mean_expression_disease  mean_expression_normal  disease_sample_size  \\\n",
      "0                    0.018547                0.022178                34225   \n",
      "1                    0.062998                0.024413                34225   \n",
      "2                    0.643530                0.533689                34225   \n",
      "3                    5.240233                7.201705                34225   \n",
      "4                    0.522891                0.544253                34225   \n",
      "...                       ...                     ...                  ...   \n",
      "1145                 0.000000                1.039532                    3   \n",
      "1146                 0.000000                0.001447                    3   \n",
      "1147                 2.496066                5.495560                    3   \n",
      "1148                 0.000000                0.197558                    3   \n",
      "1149                 0.000000                0.003274                    3   \n",
      "\n",
      "      normal_sample_size  disease_female_sample_size  \\\n",
      "0                  67753                       21797   \n",
      "1                  67753                       21797   \n",
      "2                  67753                       21797   \n",
      "3                  67753                       21797   \n",
      "4                  67753                       21797   \n",
      "...                  ...                         ...   \n",
      "1145              120480                           0   \n",
      "1146              120480                           0   \n",
      "1147              120480                           0   \n",
      "1148              120480                           0   \n",
      "1149              120480                           0   \n",
      "\n",
      "      disease_male_sample_size  normal_female_sample_size  \\\n",
      "0                        12428                      15407   \n",
      "1                        12428                      15407   \n",
      "2                        12428                      15407   \n",
      "3                        12428                      15407   \n",
      "4                        12428                      15407   \n",
      "...                        ...                        ...   \n",
      "1145                         3                        247   \n",
      "1146                         3                        247   \n",
      "1147                         3                        247   \n",
      "1148                         3                        247   \n",
      "1149                         3                        247   \n",
      "\n",
      "      normal_male_sample_size  disease_num_datasets  normal_num_datasets  \\\n",
      "0                       18492                    39                  102   \n",
      "1                       18492                    39                  102   \n",
      "2                       18492                    39                  102   \n",
      "3                       18492                    39                  102   \n",
      "4                       18492                    39                  102   \n",
      "...                       ...                   ...                  ...   \n",
      "1145                       58                     2                   64   \n",
      "1146                       58                     2                   64   \n",
      "1147                       58                     2                   64   \n",
      "1148                       58                     2                   64   \n",
      "1149                       58                     2                   64   \n",
      "\n",
      "                                       disease_datasets  \\\n",
      "0     KB1,KB2,KB4,KB6,KB7,KB8,KB10,KB14,KB15,KB16,KB...   \n",
      "1     KB1,KB2,KB4,KB6,KB7,KB8,KB10,KB14,KB15,KB16,KB...   \n",
      "2     KB1,KB2,KB4,KB6,KB7,KB8,KB10,KB14,KB15,KB16,KB...   \n",
      "3     KB1,KB2,KB4,KB6,KB7,KB8,KB10,KB14,KB15,KB16,KB...   \n",
      "4     KB1,KB2,KB4,KB6,KB7,KB8,KB10,KB14,KB15,KB16,KB...   \n",
      "...                                                 ...   \n",
      "1145                                          KB20,KB38   \n",
      "1146                                          KB20,KB38   \n",
      "1147                                          KB20,KB38   \n",
      "1148                                          KB20,KB38   \n",
      "1149                                          KB20,KB38   \n",
      "\n",
      "                                        normal_datasets  \n",
      "0     KC33,KC34,KC35,KC36,PA1,KC39,KC40,KM69,KM71,KM...  \n",
      "1     KC33,KC34,KC35,KC36,PA1,KC39,KC40,KM69,KM71,KM...  \n",
      "2     KC33,KC34,KC35,KC36,PA1,KC39,KC40,KM69,KM71,KM...  \n",
      "3     KC33,KC34,KC35,KC36,PA1,KC39,KC40,KM69,KM71,KM...  \n",
      "4     KC33,KC34,KC35,KC36,PA1,KC39,KC40,KM69,KM71,KM...  \n",
      "...                                                 ...  \n",
      "1145  KC33,KC36,PA1,KC40,KM103,PA3,PA4,PA5,HBM522.VF...  \n",
      "1146  KC33,KC36,PA1,KC40,KM103,PA3,PA4,PA5,HBM522.VF...  \n",
      "1147  KC33,KC36,PA1,KC40,KM103,PA3,PA4,PA5,HBM522.VF...  \n",
      "1148  KC33,KC36,PA1,KC40,KM103,PA3,PA4,PA5,HBM522.VF...  \n",
      "1149  KC33,KC36,PA1,KC40,KM103,PA3,PA4,PA5,HBM522.VF...  \n",
      "\n",
      "[1150 rows x 15 columns]\n"
     ]
    }
   ],
   "source": [
    "cell_types_list=list(adata_concat_scvi.obs['general_cl_label'].unique())\n",
    "marker_genes=[        \"CCL21\",\n",
    "    \"NRXN1\",\n",
    "    \"SLC9A3\",\n",
    "    \"SPP1\",\n",
    "    \"SERPINE2\",\n",
    "    \"IL18\",\n",
    "    \"NBAS\"\n",
    "    \"SLC9A3\",\n",
    "    \"TIMP2\",\n",
    "    \"IGFBP7\",\n",
    "    \"SUGP1\",\n",
    "    \"APOA1\",\n",
    "    \"MMP9\",\n",
    "    \"ITGAM\",\n",
    "    \"TFF3\",\n",
    "    \"B2M\",\n",
    "    \"S100A6\",\n",
    "    \"IL10\",\n",
    "    \"IL6\",\n",
    "    \"CXCL8\",\n",
    "    \"LCN2\",\n",
    "    \"HAVCR1\",\n",
    "    \"CST3\",\n",
    "    \"CCL2\",\n",
    "    \"FABP1\"]\n",
    "# adata_concat_scvi.var_names=adata_concat_scvi.var['gene_name']\n",
    "\n",
    "result = mean_expression_compare_to_normal(\n",
    "    adata_concat_scvi,\n",
    "    celltype_key=\"general_cl_label\",\n",
    "    as_id='as_id',\n",
    "    disease_key=\"disease\",\n",
    "    selected_celltypes=cell_types_list,\n",
    "    disease_list=['chronic kidney disease', 'acute kidney failure'],\n",
    "    selected_genes=marker_genes,\n",
    "    save_path=\"disease_vs_normal_mean_expression_long.csv\"\n",
    ")\n",
    "\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90b70a0d-b793-497f-ba6a-4f07ef797dfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "\n",
    "def diffexp_compare_to_normal(adata, celltype_key, disease_key,\n",
    "                              disease_list, selected_celltypes=None, selected_genes=None,\n",
    "                              save_path=None):\n",
    "    results = []\n",
    "\n",
    "    for disease in disease_list:\n",
    "        celltypes = selected_celltypes if selected_celltypes is not None else adata.obs[celltype_key].unique()\n",
    "\n",
    "        for celltype in celltypes:\n",
    "            mask = (adata.obs[celltype_key] == celltype) & adata.obs[disease_key].isin([disease, \"normal\"])\n",
    "            adata_sub = adata[mask, :]\n",
    "\n",
    "            if adata_sub.n_obs < 2:\n",
    "                continue\n",
    "\n",
    "            if selected_genes is not None:\n",
    "                adata_sub = adata_sub[:, selected_genes]\n",
    "\n",
    "            adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
    "\n",
    "            if \"normal\" not in adata_sub.obs[disease_key].unique():\n",
    "                print(f\"⚠️ Skipping {celltype} for {disease} (no 'normal' cells found)\")\n",
    "                continue\n",
    "\n",
    "            if disease not in adata_sub.obs[disease_key].unique():\n",
    "                print(f\"⚠️ Skipping {celltype} for {disease} (no {disease} cells found)\")\n",
    "                continue\n",
    "\n",
    "            sc.tl.rank_genes_groups(\n",
    "                adata_sub,\n",
    "                groupby=disease_key,\n",
    "                groups=[disease],\n",
    "                reference=\"normal\",\n",
    "                method=\"t-test\",\n",
    "                use_raw=False\n",
    "            )\n",
    "\n",
    "            res = adata_sub.uns[\"rank_genes_groups\"]\n",
    "            n_genes = len(res[\"names\"][disease])\n",
    "\n",
    "            for i in range(n_genes):\n",
    "                results.append({\n",
    "                    \"gene\": res[\"names\"][disease][i],\n",
    "                    \"disease\": disease,\n",
    "                    \"cell_type\": celltype,\n",
    "                    \"log2FC\": res[\"logfoldchanges\"][disease][i],\n",
    "                    \"pval\": res[\"pvals\"][disease][i],\n",
    "                    \"pval_adj\": res[\"pvals_adj\"][disease][i]\n",
    "                })\n",
    "\n",
    "    df_out = pd.DataFrame(results)\n",
    "\n",
    "    if save_path:\n",
    "        df_out.to_csv(save_path, index=False)\n",
    "        print(f\"✅ Differential expression results saved to {save_path}\")\n",
    "\n",
    "    return df_out\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "db4a5e2e-928a-4c92-8032-cf70c20f83c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(list_cl_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "9e5f9819-0ed5-4e56-83bf-dc83b30ab54d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/tmp/ipykernel_1229045/2401461207.py:33: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_sub.obs[disease_key] = adata_sub.obs[disease_key].astype(str)\n",
      "/home/wruopeng/.local/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Differential expression results saved to disease_vs_normal_ttest.csv\n",
      "     gene               disease  \\\n",
      "0    SPP1  acute kidney failure   \n",
      "1     B2M  acute kidney failure   \n",
      "2  IGFBP7  acute kidney failure   \n",
      "3    CST3  acute kidney failure   \n",
      "4   TIMP2  acute kidney failure   \n",
      "\n",
      "                                           cell_type    log2FC          pval  \\\n",
      "0  Kidney interstitial alternatively activated ma...  7.054031  8.593315e-14   \n",
      "1  Kidney interstitial alternatively activated ma...  6.007898  3.436574e-12   \n",
      "2  Kidney interstitial alternatively activated ma...  3.012910  2.164048e-09   \n",
      "3  Kidney interstitial alternatively activated ma...  2.416271  2.572660e-07   \n",
      "4  Kidney interstitial alternatively activated ma...  1.485941  1.396936e-05   \n",
      "\n",
      "       pval_adj  \n",
      "0  2.148329e-12  \n",
      "1  4.295717e-11  \n",
      "2  1.352530e-08  \n",
      "3  1.286330e-06  \n",
      "4  5.820568e-05  \n"
     ]
    }
   ],
   "source": [
    "df_ttest = diffexp_compare_to_normal(\n",
    "    adata_concat_scvi,\n",
    "    celltype_key=\"general_cl_label\",\n",
    "    disease_key=\"disease\",\n",
    "    disease_list=[\"acute kidney failure\", \"chronic kidney disease\"],\n",
    "    selected_celltypes=cell_types_list,\n",
    "    selected_genes=marker_genes,\n",
    "    save_path=\"disease_vs_normal_ttest.csv\"\n",
    ")\n",
    "\n",
    "print(df_ttest.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "68d2667b-afdc-45ce-891b-21a3ce79bfde",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged = pd.merge(\n",
    "    result,\n",
    "    df_ttest,\n",
    "    on=[\"gene\", \"disease\", \"cell_type\"],\n",
    "    how=\"left\"   # keep all rows from df_means, add t-test stats if available\n",
    ")\n",
    "\n",
    "df_merged.to_csv(\"merged_expression_with_ttest.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "2cc6bf8c-f28f-42a0-88ef-b7f845b618b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Input sequence provided is already in string format. No operation performed\n",
      "Input sequence provided is already in string format. No operation performed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Added Ensembl IDs. Saved as expression_with_ensembl.csv\n",
      "     gene                 disease                        cell_type  \\\n",
      "0   CCL21  chronic kidney disease  Proximal tubule epithelial cell   \n",
      "1   NRXN1  chronic kidney disease  Proximal tubule epithelial cell   \n",
      "2  SLC9A3  chronic kidney disease  Proximal tubule epithelial cell   \n",
      "3  SLC9A3  chronic kidney disease  Proximal tubule epithelial cell   \n",
      "4    SPP1  chronic kidney disease  Proximal tubule epithelial cell   \n",
      "\n",
      "   mean_expression_disease  mean_expression_normal  disease_sample_size  \\\n",
      "0                 0.018547                0.022178                34225   \n",
      "1                 0.062998                0.024413                34225   \n",
      "2                 0.643530                0.533689                34225   \n",
      "3                 0.643530                0.533689                34225   \n",
      "4                 5.240233                7.201705                34225   \n",
      "\n",
      "   normal_sample_size  disease_female_sample_size  disease_male_sample_size  \\\n",
      "0               67753                       21797                     12428   \n",
      "1               67753                       21797                     12428   \n",
      "2               67753                       21797                     12428   \n",
      "3               67753                       21797                     12428   \n",
      "4               67753                       21797                     12428   \n",
      "\n",
      "   normal_female_sample_size  ...  normal_num_datasets  \\\n",
      "0                      15407  ...                  102   \n",
      "1                      15407  ...                  102   \n",
      "2                      15407  ...                  102   \n",
      "3                      15407  ...                  102   \n",
      "4                      15407  ...                  102   \n",
      "\n",
      "                                    disease_datasets  \\\n",
      "0  KB1,KB2,KB4,KB6,KB7,KB8,KB10,KB14,KB15,KB16,KB...   \n",
      "1  KB1,KB2,KB4,KB6,KB7,KB8,KB10,KB14,KB15,KB16,KB...   \n",
      "2  KB1,KB2,KB4,KB6,KB7,KB8,KB10,KB14,KB15,KB16,KB...   \n",
      "3  KB1,KB2,KB4,KB6,KB7,KB8,KB10,KB14,KB15,KB16,KB...   \n",
      "4  KB1,KB2,KB4,KB6,KB7,KB8,KB10,KB14,KB15,KB16,KB...   \n",
      "\n",
      "                                     normal_datasets    log2FC          pval  \\\n",
      "0  KC33,KC34,KC35,KC36,PA1,KC39,KC40,KM69,KM71,KM... -0.260582  2.738841e-01   \n",
      "1  KC33,KC34,KC35,KC36,PA1,KC39,KC40,KM69,KM71,KM...  1.395673  3.770568e-14   \n",
      "2  KC33,KC34,KC35,KC36,PA1,KC39,KC40,KM69,KM71,KM...  0.356973  9.661869e-13   \n",
      "3  KC33,KC34,KC35,KC36,PA1,KC39,KC40,KM69,KM71,KM...  0.356973  9.661869e-13   \n",
      "4  KC33,KC34,KC35,KC36,PA1,KC39,KC40,KM69,KM71,KM... -2.836399  1.288379e-77   \n",
      "\n",
      "       pval_adj       cl_id  disease_id ratio_of_disease_over_normal_for_mean  \\\n",
      "0  3.260525e-01  CL:0002306  HP_0012622                              0.836273   \n",
      "1  9.426419e-14  CL:0002306  HP_0012622                              2.580476   \n",
      "2  2.012889e-12  CL:0002306  HP_0012622                              1.205815   \n",
      "3  2.012889e-12  CL:0002306  HP_0012622                              1.205815   \n",
      "4  6.441893e-77  CL:0002306  HP_0012622                              0.727638   \n",
      "\n",
      "        ensembl_id  \n",
      "0  ENSG00000137077  \n",
      "1  ENSG00000179915  \n",
      "2  ENSG00000281861  \n",
      "3  ENSG00000281861  \n",
      "4  ENSG00000118785  \n",
      "\n",
      "[5 rows x 22 columns]\n"
     ]
    }
   ],
   "source": [
    "import mygene\n",
    "\n",
    "import pandas as pd\n",
    "import mygene\n",
    "\n",
    "\n",
    "\n",
    "mg = mygene.MyGeneInfo()\n",
    "\n",
    "genes = df_merged[\"gene\"].unique().tolist()\n",
    "\n",
    "# Batch query mygene\n",
    "results = mg.querymany(\n",
    "    genes,\n",
    "    scopes=\"symbol\",\n",
    "    fields=\"ensembl.gene\",\n",
    "    species=\"human\"\n",
    ")\n",
    "\n",
    "res_df = pd.DataFrame(results)[[\"query\", \"ensembl\"]]\n",
    "\n",
    "def extract_ensembl(x):\n",
    "    if isinstance(x, list):\n",
    "        return x[0].get(\"gene\")\n",
    "    elif isinstance(x, dict):\n",
    "        return x.get(\"gene\")\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "res_df[\"ensembl_id\"] = res_df[\"ensembl\"].apply(extract_ensembl)\n",
    "res_df = res_df[[\"query\", \"ensembl_id\"]].rename(columns={\"query\": \"gene\"})\n",
    "\n",
    "df_merged = df_merged.merge(res_df, on=\"gene\", how=\"left\")\n",
    "\n",
    "df_merged.to_csv(\"expression_with_ensembl.csv\", index=False)\n",
    "\n",
    "print(\"✅ Added Ensembl IDs. Saved as expression_with_ensembl.csv\")\n",
    "print(df_merged.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "728ccddc-c076-44b6-878c-e816fb5e592b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Kidney interstitial alternatively activated macrophage': 'CL:1000695', 'Distal convoluted tubule epithelial cell': 'CL:0002305', 'Proximal tubule epithelial cell': 'CL:0002306', 'T cell': 'CL:0000084', 'kidney loop of Henle thick ascending limb epithelial cell': 'CL:1001106', 'Collecting duct intercalated cell': 'CL:4030005', 'Kidney interstitial fibroblast': 'CL:1000692', 'Connecting tubule epithelial cell': 'CL:1000768', 'Collecting duct principal cell': 'CL:1000714', 'Endothelial cell': 'CL:0000115', 'B cell': 'CL:0000236', 'Mononuclear phagocyte': 'CL:0000113', 'Podocyte': 'CL:0000653', 'Parietal epithelial cell': 'CL:1000452', 'Mast cell': 'CL:0000097', 'Dendritic cell': 'CL:0000451', 'Plasma cell': 'CL:0000786', 'Monocyte': 'CL:0000576', 'Loop of Henle epithelial cell ': 'CL:1001016', 'Renal interstitial pericyte': 'CL:1001318', 'Neutrophil': 'CL:0000775', 'Papillary tips cell': 'CL:1000597', 'neural cell': 'CL:0002319'}\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "def get_cl_id(cell_type_name):\n",
    "    url = \"https://www.ebi.ac.uk/ols/api/search\"\n",
    "    params = {\n",
    "        \"q\": cell_type_name,\n",
    "        \"ontology\": \"cl\",     \n",
    "        \"exact\": \"true\"        \n",
    "    }\n",
    "    response = requests.get(url, params=params)\n",
    "    if response.status_code == 200:\n",
    "        data = response.json()\n",
    "        if data['response']['numFound'] > 0:\n",
    "            hit = data['response']['docs'][0]\n",
    "            return hit.get(\"obo_id\")   # return just the CL ID\n",
    "        else:\n",
    "            return None\n",
    "    else:\n",
    "        response.raise_for_status()\n",
    "\n",
    "celltype_to_cl = {}\n",
    "\n",
    "for cell_type in cell_types_list:\n",
    "    cl_id = get_cl_id(cell_type)\n",
    "    celltype_to_cl[cell_type] = cl_id if cl_id is not None else \"NA\"\n",
    "\n",
    "print(celltype_to_cl)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "983cf0c5-7b8e-42bd-9186-2e4cc97180b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "disease_mapping={\"acute kidney failure\": \"HP_0001919\",\"chronic kidney disease\":\"HP_0012622\", \"normal\":\"unknow\"}\n",
    "\n",
    "df_merged = df_merged.rename(columns={\"mean_expression\": \"mean_expression_disease\"})\n",
    "df_merged[\"cl_id\"] = df_merged[\"cell_type\"].map(celltype_to_cl)\n",
    "df_merged[\"disease_id\"] = df_merged[\"disease\"].map(disease_mapping)\n",
    "df_merged[\"ratio_of_disease_over_normal_for_mean\"] = (\n",
    "    df_merged[\"mean_expression_disease\"] / df_merged[\"mean_expression_normal\"]\n",
    ")\n",
    "\n",
    "final_columns = [\n",
    "    \"disease\",\n",
    "    \"disease_id\",\n",
    "    \"cell_type\",\n",
    "    \"cl_id\",\n",
    "    \"gene\",\n",
    "    \"ensembl_id\",\n",
    "    \"mean_expression_disease\",\n",
    "    \"disease_sample_size\",\n",
    "    \"mean_expression_normal\",\n",
    "    \"normal_sample_size\",\n",
    "    \"disease_female_sample_size\",\n",
    "    \"disease_male_sample_size\",\n",
    "    \"normal_female_sample_size\",\n",
    "    \"normal_male_sample_size\",\n",
    "    \"disease_num_datasets\",\n",
    "    \"normal_num_datasets\",\n",
    "    \"disease_datasets\",\n",
    "    \"normal_datasets\",\n",
    "    \"ratio_of_disease_over_normal_for_mean\",\n",
    "    \"log2FC\",\n",
    "    \"pval_adj\"\n",
    "]\n",
    "\n",
    "df_final = df_merged[final_columns]\n",
    "\n",
    "df_final.to_csv(\"expression_reordered_final.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "9bfbd309-c0d4-4b45-87bc-a7a800d870ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_final = pd.merge(\n",
    "    df_final,\n",
    "    df_UBERON,\n",
    "    on=[\"cell_type\"],\n",
    "    how=\"left\"   # keep all rows from df_means, add t-test stats if available\n",
    ")\n",
    "df_final.to_csv(\"expression_reordered_final_U.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "464fab1d-fc3b-47e0-a8e1-dcbf1edf4c8f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
